holoviews/core/spaces.py

```diff
@@ -324,16 +325,18 @@ def collapse(self, dimensions=None, function=None, spreadfn=None, **kwargs):
 
         collapsed = groups.clone(shared_data=False)
         for key, group in groups.items():
-            group_data = [el.data for el in group]
-            args = (group_data, function, group.last.kdims)
             if hasattr(group.last, 'interface'):
-                col_data = group.type(group.table().aggregate(group.last.kdims, function, spreadfn, **kwargs))
-
+                group_data = concat(group)
+                if function:
+                    agg = group_data.aggregate(group.last.kdims, function, spreadfn, **kwargs)
+                    group_data = group.type(agg)
             else:
+                group_data = [el.data for el in group]
+                args = (group_data, function, group.last.kdims)
                 data = group.type.collapse_data(*args, **kwargs)
-                col_data = group.last.clone(data)
-            collapsed[key] = col_data
-        return collapsed if self.ndims > 1 else collapsed.last
+                group_data = group.last.clone(data)
```

Comment:
``group_data`` can be a whole load of different things at different times. Not critical but I would prefer to have something that isn't clobbered so much.